internal/gkr/bls12-377/solver_hints.go

```diff
@@ -21,121 +21,162 @@ import (
+	return func(_ *big.Int, ins, outs []*big.Int) error {
+		if len(ins) != 4 {
+			return fmt.Errorf("GetAssignmentHint expects 3 inputs: GKR sub-circuit index, wire index, instance index, and dummy dependency enforcer")
+		}
+		if !ins[0].IsUint64() || !ins[1].IsUint64() || !ins[2].IsUint64() {
+			return fmt.Errorf("all 3 non-dummy input to GetAssignmentHint must fit in uint64")
 		}
+		data := data[ins[0].Uint64()]
+		wireI := ins[1].Uint64()
+		instanceI := ins[2].Uint64()
+
+		data.assignment[wireI][instanceI].BigInt(outs[0])
 
-		for _, p := range info.Prints {
-			serializable := make([]any, len(p.Values))
-			for i, v := range p.Values {
-				if p.IsGkrVar[i] { // serializer stores uint32 in slices as uint64
-					serializable[i] = data.assignment[algo_utils.ForceUint32(v)][p.Instance].String()
-				} else {
-					serializable[i] = v
+		return nil
+	}
+}
+
+// SolveHint generate a hint that computes the assignments for all wires in a circuit instance.
+// It is intended for use in gkrapi.API.AddInstance.
+func SolveHint(data []SolvingData) hint.Hint {
+	return func(_ *big.Int, ins, outs []*big.Int) error {
+		if !ins[0].IsUint64() {
+			return fmt.Errorf("first input to GKR prove hint must be the sub-circuit index")
+		}
+		data := data[ins[0].Uint64()]
+
+		instanceI := ins[1].Uint64()
+
+		gateIns := make([]frontend.Variable, data.maxNbIn)
+		outsI := 0
+		insI := 2 // skip the first two input, which are the circuit and instance indices, respectively.
+		for wI := range data.circuit {
+			w := &data.circuit[wI]
+			if w.IsInput() { // read from provided input
+				data.assignment[wI][instanceI].SetBigInt(ins[insI])
+				insI++
+			} else {
+
+				// assemble input for gate
+				for i, inWI := range w.Inputs {
+					gateIns[i] = &data.assignment[inWI][instanceI]
 				}
+
+				data.assignment[wI][instanceI].Set(w.Gate.Evaluate(api, gateIns[:len(w.Inputs)]...).(*fr.Element))
+			}
+			if w.IsOutput() {
+				data.assignment[wI][instanceI].BigInt(outs[outsI])
+				outsI++
 			}
-			fmt.Println(serializable...)
 		}
 
-		setOuts(data.assignment, info.Circuit, outs)
-
 		return nil
 	}
 }
 
-func ProveHint(hashName string, data *SolvingData) hint.Hint {
+// ProveHint generates a hint that produces the GKR proof using the computed assignments contained in data.
+// It is meant for use in gkrapi.Circuit.finalize.
+func ProveHint(data []SolvingData) hint.Hint {
 
 	return func(_ *big.Int, ins, outs []*big.Int) error {
-		insBytes := algo_utils.Map(ins[1:], func(i *big.Int) []byte { // the first input is dummy, just to ensure the solver's work is done before the prover is called
+		if !ins[0].IsUint64() {
+			return fmt.Errorf("first input to GKR prove hint must be the sub-circuit index")
+		}
+		data := data[ins[0].Uint64()]
+		hashName := data.hashName
+		ins = ins[1:]
+
+		data.assignment.repeatUntilEnd(data.nbInstances)
+
+		// The second input is dummy, just to ensure the solver's work is done before the prover is called.
```

Comment:
I don't understand the comment. We are not doing any enforcement on `ins[1]` so that it is set? So how it works (my understanding):
* we call res := api.NewHint(SolveHint, circID, ...)
* we call res := api.NewHint(ProveHint, circID, res[1], .....)

Because now calling `ProveHint` depends on the outputs of calling `SolveHint`, then the gnark solver automatically orders the calling of hints so that gkr Prove is called after gkr Solve?

I get it and it works, but it most definitely needs to be written down. It really is implicit way to enforce ordering of hints. People don't understand it without knowing how gnark really works.

Reply:
Yes essentially. I'll try to improve the comment.